src/full_client.toit

```diff
@@ -221,7 +221,95 @@ interface ReconnectionStrategy:
 A base class for reconnection strategies.
 */
 abstract class DefaultReconnectionStrategyBase implements ReconnectionStrategy:
-  UNIMPLEMENTED ::= "UNIMPLEMENTED"
+  static DEFAULT_RECEIVE_CONNECT_TIMEOUT /Duration ::= Duration --s=5
+  static DEFAULT_ATTEMPT_DELAYS /List ::= [
+    Duration --s=1,
+    Duration --s=5,
+    Duration --s=15,
+  ]
+
+  receive_connect_timeout_ /Duration
+  attempt_delays_ /List
+
+  is_closed_ := false
+
+  /**
+  A latch that is set when the client is sleeping before reconnecting.
+  */
+  waiting_for_reconnect_latch_ /monitor.Latch? := null
+
+  constructor
+      --receive_connect_timeout /Duration = DEFAULT_RECEIVE_CONNECT_TIMEOUT
+      --attempt_delays /List /*Duration*/ = DEFAULT_ATTEMPT_DELAYS:
+    receive_connect_timeout_ = receive_connect_timeout
+    attempt_delays_ = attempt_delays
+
+
+  /**
+  Tries to connect, potentially retrying with delays.
+
+  Returns null if the strategy was closed.
+  Returns whether the broker had a session for this client, otherwise.
+  */
+  do_connect transport/ActivityMonitoringTransport
+      [--reconnect_transport]
+      [--send_connect]
+      [--receive_connect_ack]:
+    for i := -1; i < attempt_delays_.size; i++:
+      if is_closed: return null
+      if i >= 0:
+        // TODO(florian): we want to wait for `close` calls and `sleep` at the same
+        // time. Using a task and latch for this feels very expensive.
+        assert: waiting_for_reconnect_latch_ == null
+        waiting_for_reconnect_latch_ = monitor.Latch
+        sleeping_task := null
+        try:
+          sleeping_task = task::
+            sleep attempt_delays_[i]
+            sleeping_task = null
+            if waiting_for_reconnect_latch_ and not waiting_for_reconnect_latch_.has_value:
+              waiting_for_reconnect_latch_.set true
+
+          // As soon as the sleep is over, or there is a close, we get here.
+          waiting_for_reconnect_latch_.get
+        finally:
+          waiting_for_reconnect_latch_ = null
+          if sleeping_task: sleeping_task.cancel
+
+        if is_closed: return null
+        reconnect_transport.call
+
+      did_connect := false
+
+      is_last_attempt := (i == attempt_delays_.size - 1)
+
+      catch --unwind=(: is_closed or is_last_attempt):
+        send_connect.call
+        return with_timeout receive_connect_timeout_:
+          receive_connect_ack.call
+
+      if is_closed: return null
+
+    unreachable
+
+  abstract connect -> none
+      transport/ActivityMonitoringTransport
+      --is_initial_connection /bool
+      [--reconnect_transport]
+      [--send_connect]
+      [--receive_connect_ack]
+      [--disconnect]
+
+  abstract should_try_reconnect transport/ActivityMonitoringTransport -> bool
+
+  is_closed -> bool:
+    return is_closed_
+
+  close -> none:
+    is_closed_ = true
+    if waiting_for_reconnect_latch_ and not waiting_for_reconnect_latch_.has_value:
```

Comment:
It feels a bit like a latch is the wrong abstraction here. Why not use a semaphore and just always `up` it on close and `down` it in `do_connect`?

Reply:
https://github.com/toitware/mqtt/pull/32